admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt

```diff
@@ -0,0 +1,169 @@
+import io.ktor.auth.jwt.*
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.engine.apache.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.http.auth.*
+import io.ktor.response.*
+import io.ktor.routing.*
+import kotlinx.serialization.json.*
+import org.kodein.di.DI
+import org.kodein.di.bind
+import org.kodein.di.instance
+import org.kodein.di.singleton
+import org.kodein.di.ktor.closestDI
+import java.net.URI
+import java.net.URL
+import java.util.concurrent.TimeUnit
+
+const val JWT_COOKIE = "jwt"
+
+
+val oauthDIModule = DI.Module("oauth") {
+    configureOAuthDI()
+}
+
+fun DI.Builder.configureOAuthDI() {
+    bind<HttpClient>("oauthHttpClient") with singleton { HttpClient(Apache) }
+    bind<OAuthConfig>() with singleton { OAuthConfig(di) }
+    bind<JwkProvider>() with singleton { buildJwkProvider(instance()) }
+}
+
+fun Authentication.Configuration.configureOAuthAuthentication(di: DI) {
+    val oauthConfig by di.instance<OAuthConfig>()
+    val httpClient by di.instance<HttpClient>("oauthHttpClient")
+    val jwkProvider by di.instance<JwkProvider>()
+
+    jwt("jwt") {
+        realm = "Access to the http(s) and the ws(s) services"
+        verifier(jwkProvider, oauthConfig.issuer)
+        validate { credential ->
+            credential.toPrincipal()
+        }
+        authHeader { call ->
+            val headerValue =
+                call.request.headers[HttpHeaders.Authorization] ?: "Bearer ${call.request.cookies[JWT_COOKIE]}"
+            parseAuthorizationHeader(headerValue)
+        }
+    }
+
+    oauth("oauth") {
+        urlProvider = { URI(oauthConfig.uiRootUrl).resolve("/oauth/callback").toString() }
+        providerLookup = {
+            OAuthServerSettings.OAuth2ServerSettings(
+                name = "oauth2",
+                authorizeUrl = oauthConfig.authorizeUrl,
+                accessTokenUrl = oauthConfig.accessTokenUrl,
+                requestMethod = HttpMethod.Post,
+                clientId = oauthConfig.clientId,
+                clientSecret = oauthConfig.clientSecret,
+                defaultScopes = oauthConfig.scopes
+            )
+        }
+        client = httpClient
+    }
+}
+
+fun Routing.configureOAuthRoutes() {
+    val oauthConfig by closestDI().instance<OAuthConfig>()
+    val httpClient by closestDI().instance<HttpClient>("oauthHttpClient")
+
+    authenticate("oauth") {
+        get("/oauth/login") {
+            // Redirects to "authorizeUrl" automatically
+        }
+        get("/oauth/callback") {
+            val oauthPrincipal: OAuthAccessTokenResponse.OAuth2? = call.principal()
+            if (oauthPrincipal == null) {
+                call.unauthorizedError()
+                return@get
+            }
+
+            val userInfo = getUserInfo(httpClient, oauthConfig.userInfoUrl, oauthPrincipal.accessToken)
+            if (userInfo == null) {
+                call.unauthorizedError()
```

Comment:
__comment 1__
[RomanDavlyatshin](https://github.com/RomanDavlyatshin) [5 days ago](https://github.com/Drill4J/admin/pull/340#discussion_r1389279980)
can't we just do throw UnauthorizedError() and use centralized error handling / status pages to handle that?

__comment 2__
@[iryabov](https://github.com/iryabov) iryabov [13 minutes ago](https://github.com/Drill4J/admin/pull/340#discussion_r1394112004)
Oauth2 routes are attached outside of the main routes and common error handlers for them may not work


__comment 3__

> are attached outside of the main routes

Hmm, I would say none [of the status pages ](https://github.com/Drill4J/admin/blob/feature/oauth-sessions-EPMDJ-10707/admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt#L53) are coupled with routes down-the-line, aren't they?

I had an impression Ktor simply "places" all status pages as a sort-of-a middleware, that is applied to all routes/requests handlers

Have you checked if that works?

Reply:
Will decide later